Add a Stdin Specification

jfly opened the pull request:
This described the Stdin Specification, and adds a new `formatter.stdin_options` config to enable treefmt to invoke formatters in "stdin mode".

jfly commented on docs/site/reference/formatter-spec.md:
```diff
@@ -60,3 +60,54 @@ outputs.
+
+When in stdin mode, the formatter:
+
+1. **MUST** print the formatted file to stdout.
```
This is a lie: it's OK for the formatter to fail to format, but it probably should (must?):

- Print an explanation *why* to stderr
- Exit non-zero

This made me realize the current formatter specification doesn't have anything to say about subprocess exit codes, but treefmt does [check the exit code of the formatter it invokes](https://github.com/numtide/treefmt/blob/v2.2.1/format/formatter.go#L107). We should probably mention this in the spec?

jfly replied:
I've updated the formatter spec in https://github.com/numtide/treefmt/pull/610.

I've updated the stdin spec with the exact same rule as well.

jfly commented on docs/site/reference/formatter-spec.md:
```diff
@@ -60,3 +60,54 @@ outputs.
 
 We expect the formatter to be reliable and not break the semantics of the formatted files.
+
+## Stdin Specification
```
Should I actually promote this to a top level file? That is, `docs/site/reference/stdin-spec.md`?

jfly commented on docs/site/reference/formatter-spec.md:
````diff
@@ -60,3 +60,54 @@ outputs.
+Example:
+
+```
+$ echo "{}" | nixfmt --stdin path/to/file.nix
````
This is a bad example, as [nixfmt doesn't actually behave this way yet](https://github.com/NixOS/nixfmt/issues/305).

Do folks know if there's an existing formatter (besides treefmt itself) that implements this spec?

brianmcgee replied:
Hmmm, then I can't say I know of any formatter that actually conforms to this yet. Sounds like `ormolu` might be close?

jfly replied:
Bummer. I'm going to leave this as is, then. Hopefully we can change nixfmt to honor this spec before we merge this.

jfly replied:
Discussed at today's Nix Formatting team meeting: we like `--stdin-filepath=<path>` for 2 reasons:

1. This is path, not a filename.
2. It's more obvious that this expects exactly 1 path, rather than `[files...]`

jfly replied:
I've updated the spec to use `--stdin-filepath`

jfly commented:
@brianmcgee @zimbatm, is there anything I can do to help move this forward?

brianmcgee commented:
It's unclear to me if this documents the current behaviour or the desired behaviour. If it's the desired behaviour, isn't there some work needing to be done in `treefmt` to support that before we publish a spec?

jfly commented:
@brianmcgee, friendly ping

brianmcgee reviewed (approved):
I agree with some of your own musings in the comments you'd left. Otherwise, it seems like a sane spec and one which should be easy enough for formatters to follow. 

I also like that this is opt-in.

brianmcgee commented:
P.S. Thanks for taking the time to flesh this out, and sorry for taking a while to get around to reviewing this.

jfly commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,63 @@
+3. **MUST NOT** write to the given path on the filesytem. It _MAY_ write to
+   temporary files elsewhere on disk, but _SHOULD_ clean them up when done.
+
+### 3. Exit nonzero on error
+
+When something goes wrong when formatting (for example, the source code is
+syntactically invalid), the formatter **MUST** exit nonzero.
+
+The formatter _SHOULD_ print useful information about what went wrong to
+stderr.
```
@infinisil points out that this is redundant with the vanilla formatter spec. Remove.

infinisil reviewed (approved):
LGTM!

zimbatm commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,51 @@
+formatting "virtual files" passed via stdin.
+
+A formatter **MUST** implement the Stdin Specification if its formatting behavior
+can depend on the name of the file being formatted.
```
this is a bit hard to read

zimbatm commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,51 @@
+  "stdin mode". In stdin mode, the formatter reads file contents from stdin
+  rather than the filesystem.
+    - The formatter _MAY_ understand `--stdin-filepath=<path>` as well, but **MUST**
+      understand the space separated variant.
```
It would make sense to outline the purpose of the `<path>` argument explicitly. Both you and I know why it's there, but maybe not the reader.

zimbatm replied:
If the path argument is not needed by the tool, it's also valid to ignore it.

jfly replied:
I've taken a stab at this. Please take a look.

zimbatm commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,51 @@
+vanilla [Formatter Specification](/reference/formatter-spec), and additionally
+satisfy the following:
+
+### 1. `--stdin-filepath` flag
```
```suggestion
### 1. `--stdin` flag
```
KISS?

zimbatm reviewed (commented):
Some minor things, but directionally it makes sense.

zimbatm commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,51 @@
+
+# Stdin Specification
+
+Formatters **MAY** also implement the Stdin Specification, which allows
+formatting "virtual files" passed via stdin.
```
How is this support expressed in the treefmt.toml config?

jfly commented on docs/site/reference/stdin-spec.md:
```diff
@@ -0,0 +1,55 @@
+      directories. If the formatter's behavior doesn't depend on the given
+      `<path>`, it's ok to ignore it.
+    - The formatter _MAY_ understand `--stdin-filepath=<path>` as well, but **MUST**
+      understand the space separated variant.
```
> According to that documentation, any go-based formatter cannot support this without swapping away from the standard library's flags package

Sorry, I don't follow this point. From [the page](https://pkg.go.dev/flag#hdr-Command_line_flag_syntax) you linked to, I see this:

> The following forms are permitted:
> 
> ```
> --flag   // double dashes are also permitted
> ```

What am I missing?

> Basically please don't make me do that. [...] it's worth being flexible so this can be useful without waiting 2 years for a bunch of CLI flag adjustments to go through.

I hear you. We will definitely need a compatibility story, but it's TBD if that lives in treefmt itself, or treefmt-nix.

> If the tool does support stdin mode, I don't see why you would ever want to run it in non-stdin mode. [...] Edit: Unless you're running multiple files per formatter instance, in which case rustfmt does not duplicate work and will keep its code in cache better

This is exactly why: invoking the formatter N / B times (where B is our batch size) rather than N times.

> you still may not want this if some other formatter is single threaded

This is a good question. We probably kind of assume that formatters are single threaded.

jfly committed:
feat: add `formatter.stdin-options` to invoke formatters in stdin mode

jfly committed:
Add a Stdin Specification

This completes step 1 of
https://github.com/numtide/treefmt/issues/573#issuecomment-2877869301.

manuelteillet commented on config/config.go:
```diff
@@ -56,6 +56,12 @@ type Formatter struct {
 	Command string `mapstructure:"command" toml:"command"`
 	// Options are an optional list of args to be passed to Command.
 	Options []string `mapstructure:"options,omitempty" toml:"options,omitempty"`
+	// StringOptions are an optional list of args used to invoke the formatter
```
StringOptions -> StdinOptions

jfly replied:
Thank you, fixed.

jfly committed:
fixup! Add a Stdin Specification

jbms commented:
This addresses a pretty significant limitation in treefmt --- is there something holding up this change?  It looks pretty reasonable to me.

jfly commented:
I'm happy with the state of this, just need approval from @brianmcgee.

brianmcgee commented:
Needs a rebase, but I've approved the PR as I think this is a sane approach. Thanks for driving this, I've not had much time for treefmt of late.

jfly commented on cmd/root_test.go:
```diff
@@ -2288,12 +2288,44 @@ help:
 	// will restore the working directory during test cleanup.
 	t.Chdir("go")
+
+	contents = `
```
Seems unrelated, but this test relied upon the value of `contents` from a previous test (which was shadowed by the test I added). Rather than carefully shuffle things around to keep them happy, I opted to specify `contents` explictly. IMO, this results in a better test.

jfly commented on cmd/root_test.go:
```diff
@@ -2823,6 +2855,14 @@ func treefmt(
 		log.SetOutput(stderr)
 	}()
 
+	// treefmt may change the configured log level when we invoke it. Be sure
```
Discovered this when I added my test which calls `treefmt -vvv`, which affected the subsequent test.

jfly commented:
Thanks @brianmcgee! I've rebased, and did another review over the code. I opted to clarify the terminology we've invented here (specifically "virtual file" and "advisory path"), and I've updated the code and docs to use the terms consistently).

PTAL!
